src/corehost/src/tpafile.cpp

```diff
@@ -57,195 +75,188 @@ bool read_field(pal::string_t line, int& offset, pal::string_t& value_recv)
+            {
+                return false;
+            }
+        }
+        m_tpa_entries.push_back(entry);
     }
-
     return true;
 }
 
-void tpafile::add_from_local_dir(const pal::string_t& dir)
+void tpafile::get_local_assemblies(const pal::string_t& dir)
 {
     trace::verbose(_X("adding files from %s to TPA"), dir.c_str());
-    const pal::char_t * const tpa_extensions[] = {
-        _X(".ni.dll"),      // Probe for .ni.dll first so that it's preferred if ni and il coexist in the same dir
-        _X(".dll"),
-        _X(".ni.exe"),
-        _X(".exe"),
-    };
-
-    std::set<pal::string_t> added_assemblies;
-
-    // Get directory entries
-    auto files = pal::readdir(dir);
-    for (auto ext : tpa_extensions)
+
+    const pal::string_t tpa_extensions[] = { _X(".ni.dll"), _X(".dll"), _X(".ni.exe"), _X(".exe") };
+
+    std::vector<pal::string_t> files;
+    pal::readdir(dir, &files);
+
+    for (const auto& ext : tpa_extensions)
     {
-        auto len = pal::strlen(ext);
-        for (auto file : files)
+        for (const auto& file : files)
         {
-            // Can't be a match if it's the same length as the extension :)
-            if (file.length() > len)
+            if (file.length() <= ext.length())
             {
-                // Extract the same amount of text from the end of file name
-                auto file_ext = file.substr(file.length() - len, len);
-
-                // Check if this file name matches
-                if (pal::strcasecmp(ext, file_ext.c_str()) == 0)
-                {
-                    // Get the assembly name by stripping the extension
-                    // and add it to the set so we can de-dupe
-                    auto asm_name = file.substr(0, file.length() - len);
-
-                    // TODO(anurse): Also check if already in TPA file
-                    if (added_assemblies.find(asm_name) == added_assemblies.end())
-                    {
-                        added_assemblies.insert(asm_name);
-
-                        tpaentry_t entry;
-                        entry.asset_type = pal::string_t(_X("runtime"));
-                        entry.library_name = pal::string_t(asm_name);
-                        entry.library_version = pal::string_t(_X(""));
-
-                        pal::string_t relpath(dir);
-                        relpath.push_back(DIR_SEPARATOR);
-                        relpath.append(file);
-                        entry.relative_path = relpath;
-                        entry.asset_name = asm_name;
-
-                        trace::verbose(_X("adding %s to TPA list from %s"), asm_name.c_str(), relpath.c_str());
-                        m_entries.push_back(entry);
-                    }
-                }
+                continue;
             }
+
+            auto file_name = file.substr(0, file.length() - ext.length());
+            if (file_name.empty() || m_local_assemblies.count(file_name))
+            {
+                continue;
```

Comment:
We should add a trace entry here to callout the assembly we skipped since it was already on TPA